Misc/NEWS.d/next/Security/2022-11-04-09-29-36.gh-issue-98433.l76c5G.rst

```diff
@@ -0,0 +1,14 @@
+The IDNA codec decoder used on DNS hostnames by :mod:`socket` or :mod:`asyncio`
+related name resolution functions no longer involves a quadratic algorithm.
+This prevents a potential CPU denial of service if an out-of-spec excessive
+length hostname involving bidirectional characters were decoded. Some protocols
+such as :mod:`urllib` http ``3xx`` redirects potentially allow for an attacker
+to supply such a name.
+
+Individual labels within an IDNA encoded DNS name will now raise an error early
```

Comment:
I'm okay with leaving this length limit out either entirely - or just out of the backports if release managers are uncomfortable with this in a bugfix.

I view it as protection from other as yet unknown future idna or punycode decoder logic bugs to avoid wasting cpu time. The only reference numbers related to my arbitrary input length limit choice are: A DNS hostname is 255 bytes at most. A label within a DNS name is 63 bytes at most.

The "Nothing" characters in question that are removed in `nameprep()` in https://github.com/python/cpython/blob/main/Lib/encodings/idna.py#L18 are [non-ascii unicode oddities](https://datatracker.ietf.org/doc/html/rfc3454#section-3.1).

Other IDNA decoding libraries in the wild like the well known ICU library do not limit the length at this level of API, instead leaving that to higher levels in the application:
* https://chromium.googlesource.com/chromium/deps/icu/+/5feb9ad55c4d3609585820244d59a4462632540c/source/common/uidna.cpp#763
* https://chromium.googlesource.com/chromium/deps/icu/+/5feb9ad55c4d3609585820244d59a4462632540c/source/common/usprep.cpp#604